src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/Metadata/ModelBindingMessageProvider.cs

```diff
@@ -40,19 +40,37 @@ public abstract class ModelBindingMessageProvider
 
         /// <summary>
         /// Error message the model binding system adds when <see cref="ModelError.Exception"/> is of type
-        /// <see cref="FormatException"/> or <see cref="OverflowException"/> and value is known.
+        /// <see cref="FormatException"/> or <see cref="OverflowException"/>, value is known, and error is associated
+        /// with a property.
         /// </summary>
         /// <value>Default <see cref="string"/> is "The value '{0}' is not valid for {1}.".</value>
         public virtual Func<string, string, string> AttemptedValueIsInvalidAccessor { get; }
 
         /// <summary>
         /// Error message the model binding system adds when <see cref="ModelError.Exception"/> is of type
-        /// <see cref="FormatException"/> or <see cref="OverflowException"/> and value is unknown.
+        /// <see cref="FormatException"/> or <see cref="OverflowException"/>, value is known, and error is associated
+        /// with a collection element or action parameter.
+        /// </summary>
+        /// <value>Default <see cref="string"/> is "The value '{0}' is not valid.".</value>
+        public virtual Func<string, string> NonPropertyAttemptedValueIsInvalidAccessor { get; }
```

Comment:
> How bad would it be to just have a single `Func<string, string, string>` and put the logic in there?

Well, the resolution code is leaking into precisely two classes -- `ModelStateDictionary` and `NumericClientModelValidator`.

Making the lambda expressions more complex and different in just these cases might simplify our code. But, it would complicate what developers have to write to override one or both of our resources. That doesn't sound like a good trade-off. Let's discuss if you still feel differently.